File: android/src/main/java/au/com/crazybean/mobilex/kurir/dependency/domain/DomainModule.kt
package au.com.crazybean.mobilex.kurir.dependency.domain

import au.com.crazybean.mobilex.kurir.storage.CloudStorage
import au.com.crazybean.mobilex.kurir.impl.FirebaseStorage
import au.com.crazybean.mobilex.kurir.repository.contacts.ContactsRepository
import au.com.crazybean.mobilex.kurir.repository.contacts.ContactsSource
import au.com.crazybean.mobilex.kurir.repository.contacts.cloud.CloudContactsSource
import au.com.crazybean.mobilex.kurir.repository.geo.GeoRepository
import au.com.crazybean.mobilex.kurir.repository.geo.GeoSource
import au.com.crazybean.mobilex.kurir.repository.geo.network.GeoBytesSource
import au.com.crazybean.mobilex.kurir.repository.messages.MessagesRepository
import au.com.crazybean.mobilex.kurir.repository.messages.MessagesSource
import au.com.crazybean.mobilex.kurir.repository.messages.cloud.CloudMessagesSource
import au.com.crazybean.mobilex.kurir.repository.tasks.TasksRepository
import au.com.crazybean.mobilex.kurir.repository.tasks.TasksSource
import au.com.crazybean.mobilex.kurir.repository.tasks.cloud.CloudTasksSource
import au.com.crazybean.mobilex.kurir.repository.users.UsersRepository
import au.com.crazybean.mobilex.kurir.repository.users.UsersSource
import au.com.crazybean.mobilex.kurir.repository.users.cloud.CloudUsersSource
import org.koin.dsl.module

val domainModule = module {
    single<CloudStorage> {
        FirebaseStorage()
    }

    // Users Repository
    single<UsersSource> {
        CloudUsersSource(get())
    }
    single {
        UsersRepository(get())
    }

    // Messages Repository
    single<MessagesSource> {
        CloudMessagesSource(get())
    }
    single {
        MessagesRepository(get())
    }

    // Contacts Repository
    single<ContactsSource> {
        CloudContactsSource(get())
    }
    single {
        ContactsRepository(get())
    }

    // Tasks Repository
    single<TasksSource> {
        CloudTasksSource(get())
    }
    single {
        TasksRepository(get())
    }

    // Geo Repository
    single<GeoSource> {
        GeoBytesSource()
    }
    single {
        GeoRepository(get())
    }
}

File: ios/Kurir/dependency/domain/DomainModule.swift
//
//  DomainModule.swift
//  Kurir
//
//  Created by Loren on 19/8/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Swinject
import Mobilex

class DomainModule: Module {
    override func inject() {
        // Cloud storage
        single(CloudStorage.self) { _ in
            FirebaseStorage()
        }
        
        // Users Repository
        single(UsersSource.self) { r in
            CloudUsersSource(storage: r.resolve(CloudStorage.self)!)
        }
        
        single(UsersRepository.self) { r in
            UsersRepository(usersSource: r.resolve(UsersSource.self))
        }
        
        // Contacts Repository
        single(ContactsSource.self) { r in
            CloudContactsSource(storage: r.resolve(CloudStorage.self)!)
        }
        
        single(ContactsRepository.self) { r in
            ContactsRepository(contactsSource: r.resolve(ContactsSource.self)!)
        }
        
        // Messages Repository
        single(MessagesSource.self) { r in
            CloudMessagesSource(storage: r.resolve(CloudStorage.self)!)
        }
        
        single(MessagesRepository.self) { r in
            MessagesRepository(messagesSource: r.resolve(MessagesSource.self)!)
        }
    }
}
